### ISSUE TITLE
Piped data buffering?

### ISSUE THREAD

**Issue by aetonsi** [PowerShell/PowerShell#19036]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I'm trying to pipe a program into another. In my specific case, i'm piping an Open`ssh` console session into some "tee"ing utility (eg. `Tee-Object` or `tee.exe` from msys2, the output is exactly the same):
```powershell
ssh $env:USERNAME@localhost | tee-object log.log
# or
ssh $env:USERNAME@localhost | tee.exe log.log
```

The console output is all messed up due to (i think) some kind of buffering. It seems to me that the output is sent to the tee utility only after a newline or some particular character/byte sequence is detected. Also:
- This happens both with cmdlets and executables
- This happens even when redirecting ssh's stderr to stdout (`2>&1`)
- This does not happen at all in cmd, eg. when running ``cmd /c ssh u@h `| tee log`` (note the escaping of the pipe so that it is used by cmd instead of pwsh) or running directly in cmd
- this **might** be related to #1908 since powershell tries to interpreted redirected/piped data as strings

Please see videos below for clarity 

### Expected behavior

```console
output should appear to the console instantly, not after some particular character/byte sequence (eg. newline)
```

### Actual behavior

```console
output appears to be buffered
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.2
PSEdition                      Core
GitCommitId                    7.3.2
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0ΓÇª}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Broken powershell output
![pwsh](https://user-images.githubusercontent.com/18366087/214824308-6de59981-0c10-46c7-a496-ae71a9b720bf.gif)

### Correct command prompt output
![cmd](https://user-images.githubusercontent.com/18366087/214823338-63667438-f390-4fe7-9bbb-be051a6b8cdc.gif)

**Comment by aetonsi** [comment]
Also, if i have to invoke an expression because some argument is stored in a string variable, i can't find any way to avoid the buffering, even `iex "cmd /c ssh $sshargs user@host | tee.exe log"` is buffered and broken by powershell's `invoke-expression`...

**Comment by rhubarb-geek-nz** [comment]
PowerShell itself does not pipe byte-streams like UNIX shells or the DOS Command interpreter. Instead it uses pipes to pass an array of objects from one component to another. Piping the text output of a command does not give you an open byte-stream to read characters as they appear, instead it generally results in an array of strings.

**Comment by mklement0** [comment]
More specifically, PowerShell _streams_ output received from external programs _line by line_, decoding them into .NET strings first (if you _collect_ such output, you'll get an `[object[]]`  _array_ if there are two or more lines) and _re-encoding_ them based on `$OutputEncoding` if passed on to another external program.
In other words: You can think of this as PowerShell performing _line buffering_ (with additional overhead from decoding / re-encoding).

A related problem is that when you send (PowerShell-originated) strings _to_ an external program, a newline is invariably appended - see #5974.

> i found a hacky way 

`& cmd "/c ssh $sshargs user@host | tee.exe log"` is actually _less_ hacky than using `Invoke-Expression`, which should generally be avoided and is never needed for calling external programs.

**Comment by daxian-dbw** [comment]
The repro `ssh $env:USERNAME@localhost | tee.exe log.log` is the "_native-command piping to native-command_" scenario, which is a duplicate of https://github.com/PowerShell/PowerShell/issues/1908, which we plan to fix in 7.4 and here is the draft PR for the fix: https://github.com/PowerShell/PowerShell/pull/17857.

The repro `ssh $env:USERNAME@localhost | tee-object log.log` is the "_native_command piping to powershelll_cmdlet_" scenario. In this case, it's more of a by-design behavior.

**Comment by aetonsi** [comment]
Thanks everyone for the replies and elucidations.

@daxian-dbw just to be sure i understand, the fact that the output is "line buffered" as @mklement0 said, is all due to the pipe handling described in #1908 ? i'm asking because that discussion is (rightly) about the alteration of the bytes sequence, not about the retardation of the output handoff to the receiving end of the pipe (whether it's exe or cmdlet).

If that's the case then, simply out of curiosity, my question becomes: why is the buffering of _this particular example_ so "broken"? i thought maybe it's because of the way ssh outputs newlines but i checked and they are standard windows CRLF newlines
